tools/webidl_binder.py

```diff
@@ -243,10 +263,64 @@ def render_function(class_name, func_name, sigs, return_type, non_pointer, copy,
     body = ''
     pre_arg = []
 
+  fullname = "%s::%s" % (class_name, func_name)
+
   for i in range(max_args):
+    if i >= min_args:
+      optional = True
+    else:
+      optional = False
+    arg = allargs[i]
+    doDefault = False
+    # Filter out arguments we don't know how to parse. Fast casing only common cases.
+    compatibleArg = isinstance(arg, Dummy) or (isinstance(arg, WebIDL.IDLArgument) and arg.optional is False)
     # note: null has typeof object, but is ok to leave as is, since we are calling into asm code where null|0 = 0
-    body += "  if (arg%d && typeof arg%d === 'object') arg%d = arg%d.ptr;\n" % (i, i, i, i)
-    body += "  else arg%d = ensureString(arg%d);\n" % (i, i)
+    if not legacyMode and compatibleArg:
+      if isinstance(arg, WebIDL.IDLArgument):
+        argName = arg.identifier.name
+      else:
+        argName = ''
+      # Format assert fail message
+      checkMsg = "[CHECK FAILED] %s(arg%d:%s): " % (fullname, i, argName)
+      if isinstance(arg.type, WebIDL.IDLWrapperType):
+        inner = arg.type.inner
+      else:
+        inner = ""
+
+      # Print type info in comments.
+      body += "  /* arg%d <%s> [%s] */\n" % (i, arg.type.name, inner)
+
+      # Wrap asserts with existance check when argument is optional.
+      if allChecks and optional: body += "if(typeof arg%d !== 'undefined' && arg%d !== null) {\n" % (i, i)
+      # Special case argument types.
+      if arg.type.isNumeric():
+        if allChecks: body += "  assert(typeof arg%d === 'number' && !isNaN(arg%d), '%sExpecting <number>');\n" % (i, i, checkMsg)
```

Comment:
I disagree with these `isNaN` checks - NaN is a valid IEEE 754 value, which is why it has type `number`, and there's nothing at all nonsensical about passing it through a binding.

(If the check were restricted to integral types, then this would be awesome, but that's a different story)